### PR TITLE
[WIP] MUMUP-2874: Widgets embeddable in MyUW apps

### PR DESCRIPTION
**Pull request by thevoiceofzeke**
**In this PR**:
- Added list-of-links, option-link, rss-widget, search-with-links, custom, and basic widget types (directives and individual template files)
- Added CSS styles for all widget types
- Added a service to get widget JSON and rss feeds as JSON

*Notes:* 
- *Will add documentation about this feature in a separate PR*
- *Marked WIP while I continue to test in predev*

### Screenshots
screen shot 2017-03-28 at 12 18 04 pm
screen shot 2017-03-28 at 12 20 27 pm
screen shot 2017-03-28 at 12 22 44 pm
screen shot 2017-03-28 at 12 28 07 pm
screen shot 2017-03-28 at 12 35 12 pm

### CODE REVIEW

File: uw-frame-components/portal/widgets/controllers.js [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+      .catch(function() {
+        $log.error('Something went wrong in WidgetController');
+      });
+
+    // LOCAL FUNCTIONS
+    vm.widgetType = function widgetType(widget) {
+      var type = widget.widgetType;
+
+      if (type) {
+        if (type === 'option-link') {
```
Comment (ChristianMurphy) [inline uw-frame-components/portal/widgets/controllers.js]: Would lines 226-248 be more clear as a switch statement?

File: uw-frame-components/portal/widgets/partials/list-of-links.html [hunk 2]
```diff
@@ -0,0 +1,34 @@
+<div class="widget-body list-of-links">
+
+  <!-- FOUR LINKS OR LESS -->
+  <div ng-show="config.links.length <= 4" class="list-of-links__{{config.links.length}}">
+    <circle-button ng-repeat="item in config.links"
+                   data-href="{{item.href}}"
+                   data-target="{{item.target}}"
```
Comment (ChristianMurphy) [inline uw-frame-components/portal/widgets/partials/list-of-links.html]: Could this be more secure if it included [`rel="noopener"`](https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/)?

File: uw-frame-components/portal/widgets/partials/list-of-links.html [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                   data-trunclen="24">
+    </circle-button>
+  </div>
+
+  <!-- MORE THAN FOUR LINKS -->
+  <div ng-show="config.links.length > 4" class="list-of-links__long">
+    <ul class="widget-list">
+      <li ng-repeat='item in config.links | limitTo: 7' class="link-icon">
+        <i class="fa {{item.icon}}"></i>
+        <a ng-href="{{item.href}}" target="{{item.target}}">
```
Comment (ChristianMurphy) [inline uw-frame-components/portal/widgets/partials/list-of-links.html]: Could this be more secure if it included [`rel="noopener"`](https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/)?

File: uw-frame-components/portal/widgets/partials/list-of-links.html [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+      </li>
+    </ul>
+  </div>
+
+</div>
+
+<!-- LAUNCH BUTTON -->
+<launch-button ng-if='config.launchText && widget.url'
+               data-href="{{widget.url}}"
+               data-target="{{widget.target}}"
```
Comment (ChristianMurphy) [inline uw-frame-components/portal/widgets/partials/list-of-links.html]: Could this be more secure if it included [`rel="noopener"`](https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/)?

File: uw-frame-components/portal/widgets/partials/option-link.html [hunk 5]
```diff
@@ -0,0 +1,51 @@
+<div class="widget-option-link" layout="column" layout-align="center center">
+
+  <!-- OPTION-LINK ICON -->
+  <div class="option-link-icon">
+    <a href="{{ widget.selectedUrl }}" target="_blank" rel="noopener noreferrer">
```
Comment (ChristianMurphy) [inline uw-frame-components/portal/widgets/partials/option-link.html]: Could this be more secure if it included [`rel="noopener"`](https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/)?

File: uw-frame-components/portal/widgets/partials/rss-widget.html [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    </li>
+    <li ng-if="config.showShowing && config.lim && data.items.length > config.lim" class="no-highlight">
+      <p>Showing {{ config.lim }} of {{ data.items.length }}</p>
+    </li>
+    <li ng-if="error" class="no-highlight">There was an issue loading the information, click see all for more details
+    </li>
+    <li ng-if="isEmpty" class="no-highlight">No information at this time.</li>
+  </ul>
+</div>
+<a target="{{ config.target }}"
```
Comment (ChristianMurphy) [inline uw-frame-components/portal/widgets/partials/rss-widget.html]: Could this be more secure if it included [`rel="noopener"`](https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/)?

File: uw-frame-components/portal/widgets/partials/search-with-links.html [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        <i class="fa fa-search"></i>
+      </md-button>
+    </md-input-container>
+  </form>
+  <!-- Links -->
+  <div layout="row">
+    <div class="search-with-links__{{ config.links.length }}">
+      <circle-button ng-repeat="item in config.links"
+                     data-href="{{ item.href }}"
+                     data-target="{{ item.target }}"
```
Comment (ChristianMurphy) [inline uw-frame-components/portal/widgets/partials/search-with-links.html]: Could this be more secure if it included [`rel="noopener"`](https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/)?

File: uw-frame-components/portal/widgets/partials/search-with-links.html [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+                     data-fa-icon="{{ item.icon }}"
+                     data-disabled="false"
+                     data-title="{{ item.title }}"
+                     data-trunclen="24">
+      </circle-button>
+    </div>
+  </div>
+</div>
+<a class="launch-app-button" ng-href="widget.url" target="{{ config.target }}">
```
Comment (ChristianMurphy) [inline uw-frame-components/portal/widgets/partials/search-with-links.html]: Could this be more secure if it included [`rel="noopener"`](https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/)?

File: uw-frame-components/portal/widgets/partials/widget-card.html [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        <div ng-controller="CustomWidgetController as customWidgetCtrl">
+          <div ng-if="loading" layout="row" layout-align="center center">
+            <loading-gif data-object='content' data-empty='isEmpty'></loading-gif>
+          </div>
+          <content-item ng-if="!loading"></content-item>
+        </div>
+      </div>
+
+      <div ng-switch-when="BASIC">
+        <a tabindex="-1" ng-href="{{ renderUrl() }}" target="{{ widget.target }}" class="basic-widget">
```
Comment (ChristianMurphy) [inline uw-frame-components/portal/widgets/partials/widget-card.html]: Could this be more secure if it included [`rel="noopener"`](https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/)?

File: uw-frame-components/portal/widgets/partials/widget-card.html [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+          <div class="widget-icon-container" layout="column" layout-align="center center">
+            <widget-icon></widget-icon>
+          </div>
+        </a>
+        <launch-button data-href="{{ renderUrl() }}"
+                       data-target="_blank"
```
Comment (ChristianMurphy) [inline uw-frame-components/portal/widgets/partials/widget-card.html]: Could this be more secure if it included [`rel="noopener"`](https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/)?

File: uw-frame-components/portal/widgets/controllers.js (hunk 1 above)
Comment (thevoiceofzeke) [inline uw-frame-components/portal/widgets/controllers.js]: I'll probably include this change.

File: uw-frame-components/portal/widgets/partials/list-of-links.html (hunk 2 above)
Comment (thevoiceofzeke) [inline uw-frame-components/portal/widgets/partials/list-of-links.html]: The [circle-button directive](https://github.com/UW-Madison-DoIT/uw-frame/blob/master/uw-frame-components/portal/misc/partials/circle-button.html) includes `rel="noopener noreferrer"`

File: uw-frame-components/portal/widgets/partials/list-of-links.html (hunk 4 above)
Comment (thevoiceofzeke) [inline uw-frame-components/portal/widgets/partials/list-of-links.html]: Same case as circle-button (the `rel` is included in the directive's template)

File: uw-frame-components/portal/widgets/partials/option-link.html (hunk 5 above)
Comment (thevoiceofzeke) [inline uw-frame-components/portal/widgets/partials/option-link.html]: Beginning to suspect this is an automated comment :P

Comment (ChristianMurphy) [inline uw-frame-components/portal/widgets/partials/option-link.html]: whoops, missed that noopener was already assigned to this link

File: uw-frame-components/portal/widgets/partials/search-with-links.html (hunk 7 above)
Comment (ChristianMurphy) [inline uw-frame-components/portal/widgets/partials/search-with-links.html]: Disregard

> The circle-button directive includes rel="noopener noreferrer"

File: uw-frame-components/portal/widgets/partials/widget-card.html (hunk 10 above)
Comment (ChristianMurphy) [inline uw-frame-components/portal/widgets/partials/widget-card.html]: Already resolved.

Launch button includes [rel by default](https://github.com/UW-Madison-DoIT/uw-frame/blob/master/uw-frame-components/portal/misc/partials/launch-button.html#L3)

**Comment by ChristianMurphy** [comment]
@thevoiceofzeke
Based off d9296fd it sounds like there is an additional build system running analysis checks not being run by the public Travis CI instance.
Would there be a way to port these checks over to Travis?

**Comment by thevoiceofzeke** [comment]
This is more or less blocked for the moment. An issue I should have anticipated is that ng-portal's many uses of templates, directives, and controllers with the same names are colliding with uw-frame's things and creating a lot of problems. Since refactoring ng-portal is out of scope for this task, I'm working on some short-term workaround. 

In a perfect world, we could inject dependencies conditionally and simply not use uw-frame's widget directives/controllers unless a certain APP_FLAG is set to true in the downstream project's overlay. Unfortunately, I don't know of a way to do that.

Going to do some research -- if anyone has any ideas, let me know!

**Comment by thevoiceofzeke** [comment]
Closing this PR so folks don't get notified while I work. I'll open a new one after some decisions are made about how to proceed!

**Pull request closed without merging**